ethereum/permissioning/src/main/java/tech/pegasys/pantheon/ethereum/permissioning/node/provider/SyncStatusNodePermissioningProvider.java

```diff
@@ -0,0 +1,82 @@
+/*
+ * Copyright 2019 ConsenSys AG.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+package tech.pegasys.pantheon.ethereum.permissioning.node.provider;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import tech.pegasys.pantheon.ethereum.core.SyncStatus;
+import tech.pegasys.pantheon.ethereum.core.Synchronizer;
+import tech.pegasys.pantheon.ethereum.permissioning.node.NodePermissioningProvider;
+import tech.pegasys.pantheon.util.enode.EnodeURL;
+
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.Optional;
+
+public class SyncStatusNodePermissioningProvider implements NodePermissioningProvider {
+
+  private final Synchronizer synchronizer;
+  private final Collection<EnodeURL> bootnodes = new HashSet<>();
+  private final long syncStatusObserverId;
+  private boolean hasReachedSync = false;
+  private Optional<Runnable> hasReachedSyncCallback = Optional.empty();
+
+  public SyncStatusNodePermissioningProvider(
+      final Synchronizer synchronizer, final Collection<EnodeURL> bootnodes) {
+    checkNotNull(synchronizer);
+    this.synchronizer = synchronizer;
+    this.syncStatusObserverId = this.synchronizer.observeSyncStatus(this::handleSyncStatusUpdate);
+    this.bootnodes.addAll(bootnodes);
+  }
+
+  private void handleSyncStatusUpdate(final SyncStatus syncStatus) {
+    if (syncStatus != null && !hasReachedSync) {
+      long blocksBehind = syncStatus.getHighestBlock() - syncStatus.getCurrentBlock();
+      if (blocksBehind <= 0) {
+        synchronized (this) {
+          hasReachedSync = true;
+          hasReachedSyncCallback.ifPresent(Runnable::run);
+          synchronizer.removeObserver(syncStatusObserverId);
+        }
+      }
+    }
+  }
+
+  public boolean hasReachedSync() {
+    return hasReachedSync;
+  }
+
+  public void setHasReachedSyncCallback(final Runnable runnable) {
+    this.hasReachedSyncCallback = Optional.of(runnable);
```

Comment:
Feels like this needs to be mindful of hasReachedSync and not take the runnable if we've already reached sync or execute it at the point or something. 
Maybe as per our discussions on `NodePermissioningController` the `hasReachedSync` could be an `AtomicBoolean` and we use that to synchronise whether we're still waiting or whether we are still waiting? mild problem there where isPermitted can be blocked behind the callback getting completed. 
Maybe treat `hasReachedSyncCallback` same way as `NodePermissioningController.peerDiscoveryCallback` to prevent that runnable from getting run twice? 
Or maybe this ones just better off with more `synchronized`.